README.md
## State management, as simple as it gets

[Flutter Sample and code on Dartpad](https://dartpad.dev/?id=852bb434276d46775f61d74ece554209)

Bloobit is a state management approach inspired by the BloC pattern, but without streams (or any [observer pattern](https://en.wikipedia.org/wiki/Observer_pattern)) by default. Bloobit and the BloC pattern aim to separate presentation from business logic, but Bloobit is not an implementation of the Bloc pattern. You just call `setState()` when the state changes. There is no magic. There are only around [100 lines of code, depending on who's counting](https://github.com/MelbourneDeveloper/bloobit/blob/main/lib/bloobit.dart). You can read and understand the code.

If you need to follow the BloC pattern, I recommend [Cubit](https://pub.dev/documentation/bloc/latest/bloc/Cubit-class.html) in the [Bloc library](https://bloclibrary.dev/#/).

Bloobit is currently in beta but getting closer to a full release.

### Implement Business Logic
Extend [`Bloobit<TState>`](https://pub.dev/documentation/bloobit/latest/bloobit/Bloobit-class.html) and send messages or events to the `Bloobit` via the methods. Call [`setState`](https://pub.dev/documentation/bloobit/latest/bloobit/Bloobit/setState.html) when the state changes. 

```dart
///This extends `Bloobit` and implements the business logic with methods.
///When the state changes, we call `setState()`
class AppBloobit extends Bloobit<AppState> {
  int get callCount => state.callCount;
  bool get isProcessing => state.isProcessing;
  bool get displayWidgets => state.displayWidgets;

  final CountServerService countServerService;

  AppBloobit(this.countServerService, {void Function(AppState)? onSetState})
      : super(const AppState(0, false, true), onSetState: onSetState);

  void hideWidgets() {
    setState(state.copyWith(displayWidgets: false));
  }

  Future<void> callGetCount() async {
    setState(state.copyWith(isProcessing: true));

    final callCount = await countServerService.getCallCount();

    setState(state.copyWith(isProcessing: false, callCount: callCount));
  }
}
```

### State
We usually use immutable state, but there is no reason you can't use mutable state. `Bloobit` is agnostic about this. `Bloobit` is just a wrapper around `setState()`.

```dart
///The immutable state of the app
@immutable
class AppState {
  final int callCount;
  final bool isProcessing;
  final bool displayWidgets;

  const AppState(
    this.callCount,
    this.isProcessing,
    this.displayWidgets,
  );

  AppState copyWith({
    int? callCount,
    bool? isProcessing,
    bool? displayWidgets,
  }) =>
      AppState(
        callCount ?? this.callCount,
        isProcessing ?? this.isProcessing,
        displayWidgets ?? this.displayWidgets,
      );
}
```

You can easily inspect the state and the `Bloobit` in the widget tree using the dev tools. See the next section about the [`BloobitPropagator`](https://pub.dev/documentation/bloobit/latest/bloobit/BloobitPropagator-class.html)
![dev tools](https://github.com/MelbourneDeveloper/bloobit/blob/main/images/widgettreestate.png)

### BloobitWidget
This is the easiest way to use bloobit. Put a `BloobitWidget` in the tree and instantiate your bloobit like this. Check out the second example to see how to use bloobit with dependency injection and [ioc_container](https://pub.dev/packages/ioc_container).

```dart
class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) => MaterialApp(
        title: 'Bloobit Sample',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: BloobitWidget<AppBloobit>(
          bloobit: AppBloobit(),
          builder: (context, bloobit) => MyHomePage(
            title: 'Bloobit Sample',
            bloobit: bloobit,
          ),
        ),
      );
}
```

### Expose Bloobit Throughout the Widget Tree
If you want to nest widgets, you need to wrap your widgets in a `BloobitPropagator`. This is an [InheritedWidget](https://api.flutter.dev/flutter/widgets/InheritedWidget-class.html). The `BloobitPropagator` will pass the Bloobit to the children. 

```dart
class MyApp extends StatelessWidget {
  final IocContainer container;

  const MyApp(this.container, {super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: BloobitPropagator<AppBloobit>(
        bloobit: container.get<AppBloobit>(),
        child: const Home(),
      ),
    );
  }
}
```

### Accessing the Bloobit
You can access the `Bloobit` from any widget in the widget tree under the BloobitPropagator. Use [`BloobitPropagator.of`](https://pub.dev/documentation/bloobit/latest/bloobit/BloobitPropagator/of.html).

```dart
class CounterDisplay extends StatelessWidget {
  const CounterDisplay({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final bloobit = BloobitPropagator.of<AppBloobit>(context).bloobit;
    return Padding(
      padding: const EdgeInsets.all(10),
      child: Container(
        height: 200,
        width: 200,
        color: const Color(0xFFEEEEEE),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            if (bloobit.isProcessing)
              const CircularProgressIndicator()
            else
              Text(
                '${bloobit.callCount}',
                style: Theme.of(context).textTheme.headline4,
              ),
          ],
        ),
      ),
    );
  }
}
```

### Convert To a Stream
Bloobit is designed to work with only one `StatefulWidget`. As you can see in the example, you don't mutliple `StatefulWidgets` to power multiple child widgets. Bloc is designed to work with many, but most of the time, it is not necessary to work with multiple `StatefulWidget's. However, you can stream state changes from the `Bloobit` and use the [ioc_container](https://pub.dev/packages/ioc_container) package to wire up streaming for multiple listeners. See the example folder for the streaming code. You could use this with [StreamBuilder](https://api.flutter.dev/flutter/widgets/StreamBuilder-class.html) or multiple other `StatefulWidget's.

```dart
void main() {
  //Register services and the view model with an IoC container
  final builder = IocContainerBuilder()
    //A simple singleton service to emulate a server counting calls
    ..addSingletonService(CountServerService())
    //Adds a Stream for AppState changes
    ..addStream<AppState>()
    //The Bloobit
    ..addSingleton((con) => AppBloobit(con.get<CountServerService>(),
        onSetState: (s) =>
            //Streams state changes to the AppState stream
            con.get<StreamController<AppState>>().add(s)));

  var container = builder.toContainer();

  container
      .get<Stream<AppState>>()
      //Stream the state changes to the debug console
      .listen((appState) => debugPrint(appState.callCount.toString()));

  runApp(MyApp(container));
}
```



CHANGELOG.md
## 0.0.1-beta
Add the stuff
## 0.0.2-beta
Update the readme and example
## 0.0.3-beta
Analysis options and readme
## 0.0.4-beta
Doco fixes
## 1.0.1-beta
Rename `emit` to `setState`
## 1.0.2-beta
Documentation fixes
## 1.0.3-beta
Don't call `setState` if `State` is not mounted
## 1.0.4-beta
Add `BloobitWidget` to fill in for `StatefulWidget`
## 1.0.5-beta
Add a simpler example that uses BloobitPropagator



test/widget_test.dart
// ignore_for_file: avoid_relative_lib_imports

import 'package:bloobit/bloobit.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import '../example/lib/main.dart';
import '../example/lib/main2.dart';

void main() {
  //The traditional counter increments smoke test
  testWidgets('Counter increments smoke test', (tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });

  testWidgets(
    //This tests the example app and therefore the Bloobit library
    //It is a good example of how to test Bloobit apps
    //Nothing ties this test to Bloobit
    //You could replace the Bloobit with a Cubit and the test would still pass
    'Test The Example App',
    (tester) async {
      var streamCount = 0;

      final container = compose();

      //Listen to the stream
      //Note: streams are not a core part of Bloobit but we can attach them and
      //this test here is to ensure that onSetState is working
      container
          .get<Stream<AppState2>>()
          //Stream the state changes to the debug console
          .listen((appState) => streamCount++);

      await tester.pumpWidget(MyApp2(container));

      //Tap increment button
      await tester.tap(find.byIcon(Icons.add));
      await tester.pump(const Duration(milliseconds: 10));

      //Check that we see the progress indicators
      expect(find.byType(CircularProgressIndicator), findsNWidgets(7));

      //Settle and check the progress indicators disappeard
      await tester.pumpAndSettle(const Duration(seconds: 2));
      expect(find.byType(CircularProgressIndicator), findsNothing);

      await tester.tap(find.byIcon(Icons.close));
      await tester.pumpAndSettle();

      //Two set states on increment and one on hide
      expect(streamCount, 3);
    },
  );

  testWidgets(
    //This test tests the widgets and physically acccesses the Bloobit via
    //BloobitPropagator. You shouldn't do this because it couples your test to
    //the Bloobit library. But, this is an example of how easy it is if you
    //need to do this.
    'Test The Example App With Bloobit Access',
    (tester) async {
      await tester.pumpWidget(MyApp2(compose()));

      //Get the bloobit and check the state

      final bloobitPropagator = tester.widget<BloobitPropagator<AppBloobit2>>(
        find.byKey(bloobitPropagatorKey).first,
      );

      expect(bloobitPropagator.bloobit, isNotNull);
      expect(bloobitPropagator.bloobit.state.callCount, 0);

      //Tap increment button
      await tester.tap(find.byIcon(Icons.add));
      await tester.pump(const Duration(milliseconds: 10));
      //Check the state is processing
      expect(bloobitPropagator.bloobit.state.isProcessing, true);
      //Check that we see the progress indicators
      expect(find.byType(CircularProgressIndicator), findsNWidgets(7));

      //Settle and check the progress indicators disappeard
      await tester.pumpAndSettle(const Duration(seconds: 2));
      expect(find.byType(CircularProgressIndicator), findsNothing);

      //Check the new state
      expect(bloobitPropagator.bloobit.state.callCount, 1);
      expect(bloobitPropagator.bloobit.state.isProcessing, false);

      await tester.tap(find.byIcon(Icons.close));
      await tester.pumpAndSettle();

      //Check the final state
      expect(bloobitPropagator.bloobit.state.displayWidgets, false);
    },
  );
}


example/test/widget_test.dart
import 'package:bloobit/bloobit.dart';
import 'package:example/main.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets(
    //This tests the example app and therefore the Bloobit library
    //It is a good example of how to test Bloobit apps
    //Nothing ties this test to Bloobit
    //You could replace the Bloobit with a Cubit and the test would still pass
    'Test The Example App',
    (tester) async {
      var streamCount = 0;

      final container = compose();

      //Listen to the stream
      //Note: streams are not a core part of Bloobit but we can attach them and
      //this test here is to ensure that onSetState is working
      container
          .get<Stream<AppState>>()
          //Stream the state changes to the debug console
          .listen((appState) => streamCount++);

      await tester.pumpWidget(MyApp(container));

      //Tap increment button
      await tester.tap(find.byIcon(Icons.add));
      await tester.pump(const Duration(milliseconds: 10));

      //Check that we see the progress indicators
      expect(find.byType(CircularProgressIndicator), findsNWidgets(7));

      //Settle and check the progress indicators disappeard
      await tester.pumpAndSettle(const Duration(seconds: 2));
      expect(find.byType(CircularProgressIndicator), findsNothing);

      await tester.tap(find.byIcon(Icons.close));
      await tester.pumpAndSettle();

      //Two set states on increment and one on hide
      expect(streamCount, 3);
    },
  );

  testWidgets(
    //This test tests the widgets and physically acccesses the Bloobit via
    //BloobitPropagator. You shouldn't do this because it couples your test to
    //the Bloobit library. But, this is an example of how easy it is if you
    //need to do this.
    'Test The Example App With Bloobit Access',
    (tester) async {
      await tester.pumpWidget(MyApp(compose()));

      //Get the bloobit and check the state

      final bloobitPropagator = tester.widget<BloobitPropagator<AppBloobit>>(
        find.byKey(bloobitPropagatorKey).first,
      );

      expect(bloobitPropagator.bloobit, isNotNull);
      expect(bloobitPropagator.bloobit.state.callCount, 0);

      //Tap increment button
      await tester.tap(find.byIcon(Icons.add));
      await tester.pump(const Duration(milliseconds: 10));
      //Check the state is processing
      expect(bloobitPropagator.bloobit.state.isProcessing, true);
      //Check that we see the progress indicators
      expect(find.byType(CircularProgressIndicator), findsNWidgets(7));

      //Settle and check the progress indicators disappeard
      await tester.pumpAndSettle(const Duration(seconds: 2));
      expect(find.byType(CircularProgressIndicator), findsNothing);

      //Check the new state
      expect(bloobitPropagator.bloobit.state.callCount, 1);
      expect(bloobitPropagator.bloobit.state.isProcessing, false);

      await tester.tap(find.byIcon(Icons.close));
      await tester.pumpAndSettle();

      //Check the final state
      expect(bloobitPropagator.bloobit.state.displayWidgets, false);
    },
  );
}


lib/bloobit.dart
library bloobit;

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

///This faciliates separation of Business Logic and Presentation. You must
///extend it. Pass the setState function in by calling attach, or use
///[BloobitPropagator] to do it for you. The [Bloobit] will call the setState
///function when [setState] is called.
abstract class Bloobit<TState> {
  ///Creates a [Bloobit] with the initial state.
  Bloobit(this.initialState, {void Function(TState)? onSetState})
      : _state = initialState,
        onSetState = onSetState ??
            ((s)
                //ignore: no-empty-block
                {});

  void Function(VoidCallback fn)? _setState;

  ///This is the state of the Bloobit on construction
  final TState initialState;

  ///When state changes, the Bloobit calls this. Use this to feed state changes
  ///to a Stream
  final void Function(TState) onSetState;

  ///This is the current state of the Bloobit
  TState get state => _state;

  TState _state;

  ///Pass in the setState function from your state here
  // ignore: use_setters_to_change_properties
  void attach(void Function(VoidCallback fn) setState) => _setState = setState;

  ///This calls the setState() function and updates the state in the setState()
  ///callback
  void setState(TState state) {
    assert(_setState != null, 'You must call attach');

    //This is so we can hook up a stream or any other reactive listener
    onSetState(_state);

    _setState!(() {
      _state = state;
    });
  }
}

///Use with State class to automate the attach method
///Note: attach will not call setState if the state is not mounted
mixin AttachesSetState<TWidget extends StatefulWidget,
    TCallsSetState extends Bloobit<dynamic>> on State<TWidget> {
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    BloobitPropagator.of<TCallsSetState>(context)
        .bloobit
        .attach((s) => mounted ? setState(s) : null);
  }
}

///This InheritedWidget propagates the [Bloobit] down the widget tree
///much like Provider
class BloobitPropagator<T extends Bloobit<dynamic>> extends InheritedWidget {
  ///Creates and instance of BloobitPropagator
  const BloobitPropagator({
    required this.bloobit,
    required super.child,
    super.key,
  });

  ///This is the [Bloobit] that will be propagated
  final T bloobit;

  ///The state of the bloobit
  dynamic get state => bloobit.state;

  ///Gets the [Bloobit] from the widget tree
  static BloobitPropagator<T> of<T extends Bloobit<dynamic>>(
    BuildContext context,
  ) {
    final result =
        context.dependOnInheritedWidgetOfExactType<BloobitPropagator<T>>();
    assert(result != null, 'No Bloobit of type $T found in context');

    return result!;
  }

  @override
  bool updateShouldNotify(
    covariant InheritedWidget oldWidget,
  ) =>
      bloobit != (oldWidget as BloobitPropagator<T>).bloobit;

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties
      ..add(DiagnosticsProperty<T>('bloobit', bloobit))
      ..add(DiagnosticsProperty<dynamic>('state', state));
  }
}

///A stateful widget that uses [Bloobit] instead of [State]
class BloobitWidget<TBloobit extends Bloobit<dynamic>> extends StatefulWidget {
  ///Constructs a BloobitWidget
  const BloobitWidget({
    required this.bloobit,
    required this.builder,
    super.key,
  });

  ///The Bloobit for the widget
  final TBloobit bloobit;

  ///The builder for the widget. This should return a [StatelessWidget] and the
  ///widget can access the [Bloobit]
  final Widget Function(
    BuildContext context,
    TBloobit bloobit,
  ) builder;

  @override
  // ignore: no_logic_in_create_state
  State<StatefulWidget> createState() => _BloobitWidgetState<TBloobit>(
        builder,
        bloobit,
      );

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties
      ..add(DiagnosticsProperty<TBloobit>('bloobit', bloobit))
      ..add(
        DiagnosticsProperty<dynamic>('builder', builder),
      );
  }
}

class _BloobitWidgetState<TBloobit extends Bloobit<dynamic>>
    extends State<BloobitWidget<TBloobit>> {
  _BloobitWidgetState(
    this._builder,
    this._bloobit,
  ) {
    _bloobit.attach((s) => mounted ? setState(s) : null);
  }
  final TBloobit _bloobit;

  final Widget Function(
    BuildContext context,
    TBloobit bloobit,
  ) _builder;

  @override
  Widget build(BuildContext context) => _builder(
        context,
        _bloobit,
      );
}


example/lib/main.dart
import 'package:bloobit/bloobit.dart';
import 'package:flutter/material.dart';

//This is a very simple sample that users BloobitWidget. Check out main2.dart in
//the example folder for a more complex example that uses dependency injection
//and BloobitPropagator.

class AppBloobit extends Bloobit<int> {
  AppBloobit() : super(0);

  void increment() => setState(state + 1);
}

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) => MaterialApp(
        title: 'Bloobit Sample',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: BloobitWidget<AppBloobit>(
          bloobit: AppBloobit(),
          builder: (context, bloobit) => MyHomePage(
            title: 'Bloobit Sample',
            bloobit: bloobit,
          ),
        ),
      );
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({
    required this.bloobit,
    required this.title,
    super.key,
  });

  final String title;
  final AppBloobit bloobit;

  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(
          title: Text(title),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              const Text(
                'You have pushed the button this many times:',
              ),
              Text(
                '${bloobit.state}',
                style: Theme.of(context).textTheme.headline4,
              ),
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: bloobit.increment,
          tooltip: 'Increment',
          child: const Icon(Icons.add),
        ),
      );
}


example/lib/shared.dart
// ignore_for_file: diagnostic_describe_all_properties

import 'package:flutter/material.dart';

///The immutable state of the app
@immutable
class AppState {
  const AppState(
    this.callCount,
    this.isProcessing,
    this.displayWidgets,
  );
  final int callCount;
  final bool isProcessing;
  final bool displayWidgets;

  AppState copyWith({
    int? callCount,
    bool? isProcessing,
    bool? displayWidgets,
  }) =>
      AppState(
        callCount ?? this.callCount,
        isProcessing ?? this.isProcessing,
        displayWidgets ?? this.displayWidgets,
      );
}

///This emulates a server that counts the number of times it has been called.
class CountServerService {
  int _counter = 0;
  Future<int> getCallCount() =>
      Future<int>.delayed(const Duration(seconds: 1), () async {
        //If we return _counter++ the value is incorrect. Something strange here
        // ignore: join_return_with_assignment
        _counter++;
        return _counter;
      });
}

class CounterDisplay extends StatelessWidget {
  const CounterDisplay(
    this.state, {
    Key? key,
  }) : super(key: key);

  final AppState state;

  @override
  Widget build(BuildContext context) => Padding(
        padding: const EdgeInsets.all(10),
        child: Container(
          height: 200,
          width: 200,
          color: const Color(0xFFEEEEEE),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              const Text(
                'You have pushed the button this many times:',
              ),
              if (state.isProcessing)
                const CircularProgressIndicator()
              else
                Text(
                  '${state.callCount}',
                  style: Theme.of(context).textTheme.headline4,
                ),
            ],
          ),
        ),
      );
}


example/lib/main_cubit.dart
// ignore_for_file: depend_on_referenced_packages

import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';

class AppCubit extends Cubit<int> {
  AppCubit() : super(0);

  void increment() => emit(state + 1);
}

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    final appCubit = AppCubit();
    return MaterialApp(
      title: 'Cubit Sample',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: BlocBuilder<AppCubit, int>(
        bloc: appCubit,
        builder: (context, state) => MyHomePage(
          title: 'Cubit Sample',
          bloc: appCubit,
        ),
      ),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({
    required this.bloc,
    required this.title,
    super.key,
  });

  final String title;
  final AppCubit bloc;

  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(
          title: Text(title),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              const Text(
                'You have pushed the button this many times:',
              ),
              Text(
                '${bloc.state}',
                style: Theme.of(context).textTheme.headline4,
              ),
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: bloc.increment,
          tooltip: 'Increment',
          child: const Icon(Icons.add),
        ),
      );
}


example/lib/main_change_notifier.dart
import 'package:flutter/material.dart';

class AppChangeNotifier extends ChangeNotifier {
  //Notice that the state is mutable by default, but you can use an immutable
  //type if you want
  int counter = 0;

  void increment() {
    counter++;
    notifyListeners();
  }
}

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    final appChangeNotifier = AppChangeNotifier();

    return MaterialApp(
      title: 'Change Notifier Sample',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: AnimatedBuilder(
        animation: appChangeNotifier,
        builder: (context, bloobit) => MyHomePage(
          title: 'Change Notifier Sample',
          appChangeNotifier: appChangeNotifier,
        ),
      ),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({
    required this.appChangeNotifier,
    required this.title,
    super.key,
  });

  final String title;
  final AppChangeNotifier appChangeNotifier;

  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(
          title: Text(title),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              const Text(
                'You have pushed the button this many times:',
              ),
              Text(
                '${appChangeNotifier.counter}',
                style: Theme.of(context).textTheme.headline4,
              ),
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: appChangeNotifier.increment,
          tooltip: 'Increment',
          child: const Icon(Icons.add),
        ),
      );
}


example/lib/main2.dart
// ignore_for_file: prefer_const_literals_to_create_immutables,
// ignore_for_file: prefer_const_constructors

import 'dart:async';
import 'package:bloobit/bloobit.dart';
import 'package:flutter/material.dart';
import 'package:ioc_container/ioc_container.dart';

///The immutable state of the app
@immutable
class AppState2 {
  const AppState2(
    this.callCount,
    this.isProcessing,
    this.displayWidgets,
  );
  final int callCount;
  final bool isProcessing;
  final bool displayWidgets;

  AppState2 copyWith({
    int? callCount,
    bool? isProcessing,
    bool? displayWidgets,
  }) =>
      AppState2(
        callCount ?? this.callCount,
        isProcessing ?? this.isProcessing,
        displayWidgets ?? this.displayWidgets,
      );
}

///This extends `Bloobit` and implements the business logic with methods.
///When the state changes, we call `setState()`
class AppBloobit2 extends Bloobit<AppState2> {
  AppBloobit2(this.countServerService, {void Function(AppState2)? onSetState})
      : super(const AppState2(0, false, true), onSetState: onSetState);
  final CountServerService countServerService;

  void hideWidgets() {
    setState(state.copyWith(displayWidgets: false));
  }

  Future<void> callGetCount() async {
    setState(state.copyWith(isProcessing: true));

    final callCount = await countServerService.getCallCount();

    setState(state.copyWith(isProcessing: false, callCount: callCount));
  }
}

///This emulates a server that counts the number of times it has been called.
class CountServerService {
  int _counter = 0;
  Future<int> getCallCount() =>
      Future<int>.delayed(const Duration(seconds: 1), () async {
        //If we return _counter++ the value is incorrect. Something strange here
        // ignore: join_return_with_assignment
        _counter++;

        return _counter;
      });
}

///Shortcuts for setting up streaming etc.
extension IocContainerBuilderExtensions on IocContainerBuilder {
  ///Adds a StreamController for the given type and exposes its
  ///stream as a singleton
  void addStream<T>() => this
    ..addSingletonService(StreamController<T>.broadcast())
    ..addSingleton<Stream<T>>((con) => con.get<StreamController<T>>().stream);
}

// coverage:ignore-start
void main() {
  runApp(
    MyApp2(
      //Register services and the Bloobit with an IoC container
      compose(),
    ),
  );
}
// coverage:ignore-end

IocContainer compose() {
  final builder = IocContainerBuilder()
    //A simple singleton service to emulate a server counting calls
    ..addSingletonService(CountServerService())
    //Adds a Stream for AppState changes
    ..addStream<AppState2>()
    //The Bloobit
    ..addSingleton(
      (con) => AppBloobit2(
        con.get<CountServerService>(),
        onSetState: (s) =>
            //Streams state changes to the AppState stream
            con.get<StreamController<AppState2>>().add(s),
      ),
    );

  return builder.toContainer();
}

const bloobitPropagatorKey = ValueKey('BloobitPropagator');

class MyApp2 extends StatelessWidget {
  const MyApp2(this.container, {super.key});
  final IocContainer container;

  @override
  Widget build(BuildContext context) => MaterialApp(
        debugShowCheckedModeBanner: false,
        title: 'Flutter Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: BloobitPropagator<AppBloobit2>(
          key: bloobitPropagatorKey,
          bloobit: container.get<AppBloobit2>(),
          child: const Home(),
        ),
      );
}

class Home extends StatefulWidget {
  const Home({super.key});

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends
    //This is a normal State class
    State<Home>
    with
        //This automatically adds the ability to call setState() on the model
        AttachesSetState<Home, AppBloobit2> {
  @override
  Widget build(BuildContext context) {
    final bloobit = BloobitPropagator.of<AppBloobit2>(context).bloobit;

    return Scaffold(
      appBar: AppBar(
        title: const Text('Bloobit Sample'),
      ),
      body: Stack(
        children: [
          if (bloobit.state.displayWidgets)
            Wrap(
              children: [
                CounterDisplay(),
                CounterDisplay(),
                CounterDisplay(),
                CounterDisplay(),
                CounterDisplay(),
                CounterDisplay(),
                CounterDisplay(),
              ],
            )
          else
            Text('X', style: Theme.of(context).textTheme.headline1),
          Align(
            alignment: Alignment.bottomRight,
            child: Row(
              children: [
                FloatingActionButton(
                  onPressed: bloobit.callGetCount,
                  tooltip: 'Increment',
                  child: const Icon(Icons.add),
                ),
                FloatingActionButton(
                  onPressed: bloobit.hideWidgets,
                  tooltip: 'X',
                  child: const Icon(Icons.close),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class CounterDisplay extends StatelessWidget {
  const CounterDisplay({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    final bloobit = BloobitPropagator.of<AppBloobit2>(context).bloobit;

    return Padding(
      padding: const EdgeInsets.all(10),
      child: Container(
        height: 200,
        width: 200,
        color: const Color(0xFFEEEEEE),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            if (bloobit.state.isProcessing)
              const CircularProgressIndicator()
            else
              Text(
                '${bloobit.state.callCount}',
                style: Theme.of(context).textTheme.headline4,
              ),
          ],
        ),
      ),
    );
  }
}
